contracts/LeaverTokenLock.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.7;
+
+import "./TokenLock.sol";
+
+contract LeaverTokenLock is TokenLock {
+    using SafeERC20 for IERC20;
+
+    enum LeaverType {
+        None,
+        GoodLeaver,
+        BadLeaver
+    }
+
+    event LogLeave(address indexed leaver, LeaverType leaverType, uint256 newTotalStake);
+
+    // manager address that can execute leave method
+    address internal _manager;
+
+    // information on leavers
+    mapping(address => LeaverType) internal _leavers;
+
+    modifier onlyManager() {
+        require(msg.sender == _manager, "Only manager");
+        _;
+    }
+
+    modifier onlyNonLeaver(address wallet) {
+        require(_leavers[wallet] == LeaverType.None, "Specified wallet already left");
+        _;
+    }
+
+    constructor(
+        IERC20 token_,
+        uint256 startTime_,
+        uint256 vestingPeriod_,
+        address manager_,
+        address[] memory wallets_,
+        uint128[] memory amounts_
+    ) TokenLock(token_, startTime_, vestingPeriod_, 0, 0, wallets_, amounts_) {
+        _manager = manager_;
+    }
+
+    function managerAddress() public view returns (address) {
+        return _manager;
+    }
+
+    function hasLeaved(address wallet) public view returns (LeaverType) {
+        return _leavers[wallet];
+    }
+
+    function leaveWallet(address wallet, bool isBadLeaver) public onlyManager onlyNonLeaver(wallet) {
+        require(msg.sender != wallet, "Manager cannot leave");
```

Comment:
Why are you using msg.sender rather than _manager?

Reply:
msg.sender must be manager, there's a modifier `onlyManager` but I agree that check in the require is misleading and that is not a bug just due to "luck". I'll fix that.